### PR TITLE
P4 3469 vehicle reg on journey events

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
This change adds support for the storage and processing of event details, in particular around journey events with details containing registration details of the vehicle used for the journey start and complete and complete.

The change was larger than originally anticipated as the details on the events were not being persisted (they were ignored).

**Before:**

image

**After:**

image

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/controller/JourneyPriceCatalogueController.kt [hunk 1]
```diff
@@ -13,26 +13,26 @@ import org.springframework.web.bind.annotation.RestController
...
 import javax.servlet.http.HttpServletResponse
 
-private val logger = loggerFor<OutputSpreadsheetController>()
+private val logger = loggerFor<JourneyPriceCatalogueController>()
 
 @RestController
 @SessionAttributes(SUPPLIER_ATTRIBUTE)
-class OutputSpreadsheetController(
-  private val spreadsheetService: SpreadsheetService,
+class JourneyPriceCatalogueController(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/controller/JourneyPriceCatalogueController.kt:28]: Unrelated renaming change.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/domain/move/Event.kt [hunk 2]
```diff
@@ -46,8 +52,9 @@ data class Event constructor(
   @Column(name = "eventable_id")
   val eventableId: String,
 
-  @Transient
-  val details: Map<String, Any>?,
+  @Convert(converter = DetailsConverter::class)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/domain/move/Event.kt:55]: So event details (i.e. reg details) were never persisted.  I am not going to worry about backfilling these for now.  I think we are probably fine to say going forwards we will be using the details from the events.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/domain/move/MovePersister.kt [hunk 3]
```diff
@@ -120,27 +118,31 @@ class MovePersister(
...
   fun processJourneys(move: Move, journeys: List<Journey>, journeyEvents: List<Event>): List<Journey> {
     return if (move.moveType() == MoveType.CANCELLED && journeys.isEmpty())
-      listOf(fakeCancelledJourneyForPricing(move)) // add a fake cancelled journey
+      listOf(fakeCancelledJourneyForPricing(move))
     else
       journeys.map { journey ->
         val thisJourneyEvents = journeyEvents.filter { it.eventableId == journey.journeyId }
         val pickUp = Event.getLatestByType(thisJourneyEvents, EventType.JOURNEY_START)?.occurredAt
         val dropOff = Event.getLatestByType(thisJourneyEvents, EventType.JOURNEY_COMPLETE)?.occurredAt
+        val vehicleRegistration = journey.determineRegistrationsCombinedWithThoseFrom(thisJourneyEvents)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/domain/move/MovePersister.kt:127]: Events are actually transient on the journey so we apply they to copy of the journey here and work out the registration to be applied to the persistent attribute on the journey below.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/JourneyPriceCatalogueService.kt [hunk 4]
```diff
@@ -10,16 +10,16 @@ import uk.gov.justice.digital.hmpps.pecs.jpc.util.loggerFor
 import java.io.File
 import java.time.LocalDate
 
-private val logger = loggerFor<SpreadsheetService>()
+private val logger = loggerFor<JourneyPriceCatalogueService>()
 
 @Transactional
 @Service
-class SpreadsheetService(
+class JourneyPriceCatalogueService(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/JourneyPriceCatalogueService.kt:17]: Unrelated naming change.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/domain/move/Event.kt [hunk 5]
```diff
@@ -46,8 +53,10 @@ data class Event constructor(
   @Column(name = "eventable_id")
   val eventableId: String,
 
-  @Transient
-  val details: Map<String, Any>?,
+  @JsonDetailsConverter
+  @Convert(converter = JpaDetailsConverter::class)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/domain/move/Event.kt:57]: This is JPA wizardry. Essentially this gets flattened out into a JSON string representation of the details from the original event json e.g. `{"vehicle_reg": "ABCDEDF"}`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/domain/move/Event.kt [hunk 6]
```diff
@@ -106,7 +148,9 @@ data class Event constructor(
 
     fun fromJson(json: String): Event? {
       return Klaxon().fieldConverter(SupplierParser::class, supplierConverter)
-        .fieldConverter(EventDateTime::class, dateTimeConverter).parse<Event>(json)
+        .fieldConverter(EventDateTime::class, dateTimeConverter)
+        .fieldConverter(JsonDetailsConverter::class, jsonDetailsConverter)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/domain/move/Event.kt:152]: This tells it how to convert the source JSON in the Details object.

**Review by thomasleese — APPROVED** [review]
